### PR TITLE
Added doc on keyword Pub.

### PR DESCRIPTION
**Pull request by ghost**
Hi, this is my first pull request. I hope it's OK. Please let me know if it would benefit from any changes. Thank you.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
It looks like you accidentally created the file at the top-level directory, instead of editing the existing file at `src/libstd/keyword_docs.rs`. Could you try to move the new file atop the old one? Something like `mv keywords_docs.rs src/libstd/keyword_docs.rs` should do the right thing. Afterwards you'd commit and push to this branch.

Let me know if I can help further!

**Comment by sjud** [comment]
Hi,
Thank you for your feedback! I'm still getting used to git and I appreciate your patience.

File: src/libstd/keyword_docs.rs [hunk 1]
```diff
@@ -925,8 +925,42 @@ mod mut_keyword {}
 //
 /// Make an item visible to others.
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
+///`pub` makes any module, function, or data structure accessible from inside
```
Comment (Mark-Simulacrum) [inline src/libstd/keyword_docs.rs]: Please add a leading column of whitespace after the `///`, like in the other lines.

File: src/libstd/keyword_docs.rs [hunk 2]
```diff
@@ -925,8 +925,42 @@ mod mut_keyword {}
...
+/// of external modules.
+/// ```rust
+/// mod other_mod {
+///     pub mod public_module {
+///         pub struct PublicStruct;
```
Comment (Mark-Simulacrum) [inline src/libstd/keyword_docs.rs]: This seems to be missing a `}` line after it.

File: src/libstd/keyword_docs.rs [hunk 3]
```diff
@@ -925,8 +925,42 @@ mod mut_keyword {}
...
+/// }
+///     let emptyStruct = other_mod::public_module::PublicStruct;
```
Comment (Mark-Simulacrum) [inline src/libstd/keyword_docs.rs]: Let's use a `pub use` statement here, to demonstrate that syntax and also avoid the compile error from a let in item context.

We may also need to add a `# fn main() {}` to make rustdoc happy.

File: src/libstd/keyword_docs.rs [hunk 4]
```diff
@@ -925,8 +925,42 @@ mod mut_keyword {}
...
+/// ```
+/// private_module does not use the `pub` keyword and therefore
+/// HiddenStruct, despite it's usage of the `pub` is inaccessible.
+///
+/// ```rust
```
Comment (Mark-Simulacrum) [inline src/libstd/keyword_docs.rs]: We should make this test case as `rust,compile_fail` and include the compilation error uncommented.

File: src/libstd/keyword_docs.rs [hunk 5]
```diff
@@ -925,8 +925,42 @@ mod mut_keyword {}
...
+///     mod private_module {
+///         pub struct HiddenStruct;
+///     }
+/// }
+/// // let emptyStruct = other_mod::private_module::HiddenStruct; <- This will not compile.
+/// ```
+/// Likewise, code inside of a public module, is inaccessible
+/// unless it uses the `pub` keyword.
+///
+/// ```rust
```
Comment (Mark-Simulacrum) [inline src/libstd/keyword_docs.rs]: Please also make this a compile_fail doctest.

**Review by Mark-Simulacrum — COMMENTED** [review]
Looking good. Some minor nits, let me know if I can help fix them.

**Comment by sjud** [comment]
Hi, Mark.
Thanks for your feedback! I've made the changes and added some further details regarding pub use. Please let me know what you think, thank you.

File: src/libstd/keyword_docs.rs [hunk 6]
```diff
@@ -925,8 +925,81 @@ mod mut_keyword {}
...
+///
+/// // We can't use private function because it's private.
+/// private_function();
+/// ```
 ///
+/// For more information on the `pub` keyword, please see the [module] section
+/// of the Rust Book detailing paths. Or the [visibility] section of Rust By Example.
+/// [module]: ../book/ch7-3-paths-for-referring-to-an-item-in-the-module-tree.html
+/// [visibility]:../rust-by-example/mod/visibility.html
 /// [not yet complete]: https://github.com/rust-lang/rust/issues/34601
```
Comment (Mark-Simulacrum) [inline src/libstd/keyword_docs.rs]: This is no longer used, so can be dropped.

File: src/libstd/keyword_docs.rs [hunk 7]
```diff
@@ -925,8 +925,81 @@ mod mut_keyword {}
...
+///     pub mod public_module {
+///         fn private_function() -> () {}
+///     }
+/// }
+///
+/// // We can't use private function because it's private.
+/// private_function();
+/// ```
 ///
+/// For more information on the `pub` keyword, please see the [module] section
```
Comment (Mark-Simulacrum) [inline src/libstd/keyword_docs.rs]: I think linking to https://doc.rust-lang.org/reference/visibility-and-privacy.html is a better fit here, it seems like the most exhaustive and best documentation we have right now.

File: src/libstd/keyword_docs.rs [hunk 8]
```diff
@@ -925,8 +925,81 @@ mod mut_keyword {}
 //
 /// Make an item visible to others.
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
+/// The keyword `pub` makes any module, function, or data structure accessible from inside
+/// of external modules. The `pub` keyword may also be used in a `use` declaration to re-export
+/// an identifier from a namespace.
+///
+///  In the below example we define `PublicStruct` within `public_inner_module` and make both public
```
Comment (Mark-Simulacrum) [inline src/libstd/keyword_docs.rs]: In thinking about this some more, I would personally prefer to drop the code examples from this documentation and primarily lead people towards the [visibility documentation in the reference](https://doc.rust-lang.org/reference/visibility-and-privacy.html).

I think in practice that means leaving the first paragraph in, and then deleting the code examples and such up to the final paragraph. I think that this is just not the best place for long-form documentation like this.

**Comment by sjud** [comment]
Thanks, Mark.
That makes sense to me.

File: src/libstd/keyword_docs.rs [hunk 9]
```diff
@@ -925,8 +925,14 @@ mod mut_keyword {}
...
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
-///
+/// The keyword `pub` makes any module, function, or data structure accessible from inside
+/// of external modules. The `pub` keyword may also be used in a `use` declaration to re-export
+/// an identifier from a namespace.
+///
+/// For more information on the `pub` keyword, please see the [visibility] section
+/// of the Reference. Or the [visibility] section of Rust By Example.
+/// [visibility]: ../reference/visibility-and-privacy.html?highlight=pub#visibility-and-privacy
+/// [visibility]:../rust-by-example/mod/visibility.html
```
Comment (Mark-Simulacrum) [inline src/libstd/keyword_docs.rs]: These will need to be distinct identifiers; one option is to reword the above to be: `please see the visibility sections of the [Reference] and examples in the [Rust by Example] collection.`.

File: src/libstd/keyword_docs.rs [hunk 10]
```diff
@@ -925,8 +925,14 @@ mod mut_keyword {}
...
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
-///
+/// The keyword `pub` makes any module, function, or data structure accessible from inside
+/// of external modules. The `pub` keyword may also be used in a `use` declaration to re-export
+/// an identifier from a namespace.
+///
+/// For more information on the `pub` keyword, please see the [Reference] section
+/// of the Reference. Or the [Rust by Example] section of Rust By Example.
+/// [Reference]: ../reference/visibility-and-privacy.html?highlight=pub#visibility-and-privacy
```
Comment (Mark-Simulacrum) [inline src/libstd/keyword_docs.rs]:
```suggestion
/// For more information on the `pub` keyword, please see the visibility section
/// of the [reference] and for some examples, see [Rust by Example].
/// 
/// [reference]: ../reference/visibility-and-privacy.html?highlight=pub#visibility-and-privacy
```

Comment (sjud) [reply src/libstd/keyword_docs.rs]: Just by reading the error message I don't think I would have figured out that the capitalization in [reference] was the problem. Thanks for this suggestion!

File: src/libstd/keyword_docs.rs [hunk 11]
```diff
@@ -925,8 +925,15 @@ mod mut_keyword {}
...
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
-///
+/// The keyword `pub` makes any module, function, or data structure accessible from inside
+/// of external modules. The `pub` keyword may also be used in a `use` declaration to re-export
+/// an identifier from a namespace.
+///
+/// For more information on the `pub` keyword, please see the visibility section
+/// of the [reference] and for some examples, see [Rust by Example].
+///
```
Comment (Mark-Simulacrum) [inline src/libstd/keyword_docs.rs]:
```suggestion
///
```

File: src/libstd/keyword_docs.rs [hunk 12]
```diff
@@ -925,8 +925,15 @@ mod mut_keyword {}
...
+/// The keyword `pub` makes any module, function, or data structure accessible from inside
+/// of external modules. The `pub` keyword may also be used in a `use` declaration to re-export
+/// an identifier from a namespace.
+///
+/// For more information on the `pub` keyword, please see the visibility section
+/// of the [reference] and for some examples, see [Rust by Example].
+/// 
+/// [reference]: ../reference/visibility-and-privacy.html?highlight=pub#visibility-and-privacy
+/// [Rust by Example]:../rust-by-example/mod/visibility.html
 /// [not yet complete]: https://github.com/rust-lang/rust/issues/34601
```
Comment (Mark-Simulacrum) [inline src/libstd/keyword_docs.rs]:
```suggestion
```

**Review by Mark-Simulacrum — COMMENTED** [review]
I think the blank line in between was actually the problem though not sure :)

**Comment by sjud** [comment]
> I think the blank line in between was actually the problem though not sure :)

I'm not sure either! I deleted the leading white space in front of our [reference] link in case that mattered. Always exciting to see what the compiler has to say. :D

**Comment by Mark-Simulacrum** [comment]
This is looking great! I think the last thing to do is to squash the commits into one. I would recommend to use `git rebase -i origin/master` (if origin is your rust-lang/rust remote) and then change "pick" to "f" on all but the first line. Once that's done, you should have just one commit left which you can then git push -f up to this branch.

If you want I can also do this for you and push up to this branch, or alternatively feel free to ping me on Zulip or Discord (`@simulacrum`) too.

**Comment by sjud** [comment]
Hi, Mark. Thanks so much for your help!
I am not very familiar with git, and have been mostly using the browser. In my next pull request I'll definitely be using the terminal more, but if you could help me rebase this one I'd appreciate it.

**Comment by Mark-Simulacrum** [comment]
No problem!

@bors r+ rollup=always

**Pull request merged**